crates/zksync/compiler/src/zksolc/mod.rs

```diff
@@ -86,14 +90,102 @@ pub trait FindContract {
 
     /// Finds a contract matching the ZK bytecode hash
     fn find_zk_bytecode_hash(&self, code_hash: H256) -> Option<&DualCompiledContract>;
+
+    /// Finds a contract matching the ZK deployed bytecode
+    fn find_zk_deployed_bytecode(&self, bytecode: &[u8]) -> Option<&DualCompiledContract>;
+
+    /// Finds a contract own and nested factory deps
+    fn fetch_all_factory_deps(&self, root: &DualCompiledContract) -> HashSet<Vec<u8>>;
 }
 
 impl FindContract for Vec<DualCompiledContract> {
+    fn find_zk_deployed_bytecode(&self, bytecode: &[u8]) -> Option<&DualCompiledContract> {
+        self.iter().find(|contract| bytecode.starts_with(&contract.zk_deployed_bytecode))
+    }
+
     fn find_evm_bytecode(&self, bytecode: &[u8]) -> Option<&DualCompiledContract> {
         self.iter().find(|contract| bytecode.starts_with(&contract.evm_bytecode))
     }
 
     fn find_zk_bytecode_hash(&self, code_hash: H256) -> Option<&DualCompiledContract> {
         self.iter().find(|contract| code_hash == contract.zk_bytecode_hash)
     }
+
+    fn fetch_all_factory_deps(&self, root: &DualCompiledContract) -> HashSet<Vec<u8>> {
+        let mut visited = HashSet::new();
+        let mut queue = VecDeque::new();
+
+        for dep in root.zk_factory_deps.iter().cloned() {
+            queue.push_back(dep);
+        }
+
+        while let Some(dep) = queue.pop_front() {
+            //try to insert in the list of visited, if it's already present, skip
+            if visited.insert(dep.clone()) {
```

Comment:
Probably we can do this without cloning and return a references to the bytecodes, but let's leave it at that and return `Vec<Vec<u8>>` instead to be consistent with the rest of the codebase.